src/H5Dmpio.c

```diff
@@ -3620,55 +3717,95 @@ H5D__mpio_share_chunk_modification_data(H5D_filtered_collective_io_info_t *chunk
-            HGOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "couldn't allocate send requests buffer")
+        send_bufs_nalloc = H5D_CHUNK_NUM_SEND_MSGS_INIT;
+        if (NULL == (msg_send_bufs = H5MM_malloc(send_bufs_nalloc * sizeof(*msg_send_bufs))))
+            HGOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "couldn't allocate chunk modification message buffer array")
+
+        if (NULL == (send_requests = H5MM_malloc(send_bufs_nalloc * sizeof(*send_requests))))
+            HGOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "couldn't allocate send requests array")
     }
 
-    /* Perform all the sends on the chunks that this rank doesn't own.
-     * Here, the sends and recvs are posted in two separate loops to
-     * avoid a deadlock
+    /*
+     * For each chunk this rank owns, add to the total number of
+     * incoming MPI messages, then update the local chunk list to
+     * overwrite any previous chunks no longer owned by this rank.
+     * Since the data for those chunks will have already been sent,
+     * this rank should no longer be interested in them and they
+     * can effectively be discarded. This bookkeeping also makes
+     * the code for the collective file space re-allocation and
+     * chunk re-insertion operations a bit simpler.
+     *
+     * For each chunk this rank doesn't own, use non-blocking
+     * synchronous sends to send the data this rank is writing to
+     * the rank that does own the chunk.
      */
-    for (i = 0; i < *chunk_list_num_entries; i++) {
+    for (i = 0, last_assigned_idx = 0; i < *chunk_list_num_entries; i++) {
         H5D_filtered_collective_io_info_t *chunk_entry = &chunk_list[i];
 
-        if (mpi_rank != chunk_entry->new_owner) {
+        if (mpi_rank == chunk_entry->new_owner) {
+            num_msgs_incoming += (size_t)(chunk_entry->num_writers - 1);
+
+            /*
+             * Overwrite chunk entries this rank doesn't own with entries that it
+             * does own, since it has sent the necessary data and is no longer
+             * interested in the chunks it doesn't own.
+             */
+            chunk_list[last_assigned_idx] = chunk_list[i];
+
+            /*
+             * Since, at large scale, a chunk's index value may be larger than
+             * the maximum value that can be stored in an int, we cannot rely
+             * on using a chunk's index value as the tag for the MPI messages
+             * sent/received for a chunk. Therefore, add this chunk to a hash
+             * table with the chunk's index as a key so that we can quickly find
+             * the chunk when processing chunk messages that were received. The
+             * message itself will contain the chunk's index so we can update
+             * the correct chunk with the received data.
+             */
+            HASH_ADD(hh, chunk_table, index_info.chunk_idx, sizeof(hsize_t), &chunk_list[last_assigned_idx]);
```

Comment:
Yes, uthash appears to use an unsigned for the hash value, so there are going to be hash collisions here. It's maybe not ideal, but since the chunk index can't be used for the MPI message tag and messages can come in any order, I needed a different way to quickly find a particular chunk structure by chunk index without iterating through the local array each time.

If we were to convert the library's chunk info structure to also include stuff needed for parallel compression, we could probably just do a skip list search to find a chunk. That may or may not be better here on average depending on the number of hash collisions and how well collisions are resolved, but there's a decent bit more work to do there. Open to suggestions.